### ISSUE TITLE
Akka.Net should allow for "Strongly typed" actors and actor references

### ISSUE THREAD

**Issue by neils-s** [akkadotnet/akka.net#2334]
When I define an `IActorRef` in code, I have no way to constrain the type of actor that will be referenced by that variable.  This is bad because it allows me to screw up my code by building a method that expects one type of `IActorRef`, but receives another.

As an example, say I have the following code:

```
using Akka.Actor;

namespace Test_of_Akka_functionality
{
    class Driver
    {
        public class GoodActor:ReceiveActor
        {
            public GoodActor()
            {
                Receive<goodMessage>(
                    m => onGoodMessage((goodMessage)m)
                    );
            }
            public class goodMessage { }
            private void onGoodMessage(goodMessage message) { }
        }
        public class BadActor : ReceiveActor
        {
            public BadActor()
            {
                Receive<badMessage>(
                    m => onBadMessage((badMessage)m)
                    );
            }
            public class badMessage { }
            private void onBadMessage(badMessage message) { }
        }

        public static void Main(string[] args)
        {
            ActorSystem MyActorSystem = ActorSystem.Create("MyActorSystem");
            IActorRef aGoodActor = MyActorSystem.ActorOf(Props.Create<GoodActor>("a_good_little_actor"));
            IActorRef aBadActor = MyActorSystem.ActorOf(Props.Create<BadActor>("bad_actor_no_candy"));

            sendGoodMessage(aBadActor);
        }

        public static void sendGoodMessage(IActorRef whichActor)
        {
            whichActor.Tell(new GoodActor.goodMessage());
        }
    }
}
```

This code gives no compiler errors, even though I've built bad code that sends a good message to a bad actor.  Ideally, I could specify that `sendGoodMessage()` takes in an instance of `IActorRef` that refers to a _specific_ kind of actor.
For example, if Akka.Net allowed generic types like `IActorRef<GoodActor>` and `IActorRef<BadActor>`, then I could specify the function signature of sendGoodMessage as:

```
        public static void sendGoodMessage(IActorRef<GoodActor> whichActor)
        {
            whichActor.Tell(new GoodActor.goodMessage());
        }
```

This would make it harder for me to shoot myself in the foot.

There are a couple of similar complaints on Stackoverflow.com that reference GitHub repositories where people are trying to work around this deficit.  For additional details see the following:
[http://stackoverflow.com/questions/37607785/how-to-ensure-at-compile-time-that-messages-are-send-to-the-right-actor-in-akka](url)
[http://stackoverflow.com/questions/37734390/strong-typing-for-iactorref-for-dependencies)](url)

**Comment by Danthar** [comment]
Maybe https://github.com/SaladLab/Akka.Interfaced is something worth looking at ?

**Comment by neils-s** [comment]
I'm playing around with Akka.Net as part of a production project.  I get worried about including too many different repos in my dev stack.  Especially repos that are on major version "0".  My boss gets nervous about that sort of stuff.  ;-)
I'm currently coding around this issue by creating some container classes for Props and IActorRef, but it seems like this is an issue that other people are going to run into.  Also, shimming container classes into end-developer code has some 'smell' to it, so it would be nice not to rely on it in perpetuity.

**Comment by Horusiath** [comment]
@neils-s for C# there are no typed actor refs at the moment. However, in [Akkling](https://github.com/Horusiath/Akkling) you have them. 

I've created a [gist](https://gist.github.com/Horusiath/f5c77e043ee17e02d0311315bcde9c79) which moves most of that functionality from Akkling to be used in C#.

**Comment by neils-s** [comment]
Thanks, Horusiath!  I created something similar to your example (though I used classes instead of structs as containers), but I like your code better.
Hopefully, typed actor refs will get included in the master branch of Akka.Net as a future enhancement.  Until then, I think I've got enough to "strongly type" my actors and protect myself from myself.

**Comment by rogeralsing** [comment]
@Horusiath wouldn't it just be both simpler and safer to make the Typed ref not be an `IActorRef`?
If it is an `IActorRef` it can still receive other types of messages if passed to anything that is expecting `IActorRef` unless you do type assertions on messages but that would screw up auto receive messages etc.

So by having a generic wrapper, you could still pass it around, and there is no way to have it accidently receive messages of other types.

Sure, you wouldn't be able to use it with things like the scheduler or eventbus, but IMO it might be better to explicitly unwrap the wrapper in those cases.

**Comment by Horusiath** [comment]
@rogeralsing yes, you're right. Typed &rArr; Untyped actor ref conversion can be done easily with custom cast operators. My snippet was based on F#, where all interface casts must be explicit.

**Comment by MrTortoise** [comment]
I thought not having this was a conscious design decision!

**Comment by earthdmitriy** [comment]
Hello! I agree with @neils-s that we need strongly typed actors and actor refs but I have another point of view regarding reasons of it.

I'm playing with demo asp.net mvc project that utilize akka.net as business logic layer. Controllers only pass messages to actor system.

I found that:
1) It's hard to pass actorRef that represent correct actor type into controller using dependency injection.

For now it:
```
public class CRUDController : Controller
{
        public CRUDController(IActorRef crudActorRef)
        {
            this.CrudActorRef = crudActorRef;
        }
}
```
IoC container know nothing about expected type of actor. Possible solution - configure IoC container to inject into specific controller specific actorRef. It will complicate the code.

All examples that I found use one of 2 approaches:
1) ActorRefs are public static variables - good bye unit tests.
2) Into controller being injected actor system itself. In case of unit tests it's required to mock additional entity - ActorSystem. And IMO it break the single responsibility principle. Controller should do things, but not worry about getting the correct actor from actor system.

I would like to write:
```
public class CRUDController : Controller
{
        public CRUDController(IActorRef<CrudActor> crudActorRef)
        {
            this.CrudActorRef = crudActorRef;
        }
}
```
IoC container will know about required actor type and will be able to provide correct actorRef for it.

Possible implementation:
Add generic version of of IActorRefFactory ActorOf(...)
